### ISSUE TITLE
Soundness bug with GADTs and lazy

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7421]
**Original bug ID:** 7421
**Reporter:** @lpw25
**Assigned to:** @garrigue
**Status:** resolved (set by @garrigue on 2016-12-10T02:48:23Z)
**Resolution:** fixed
**Priority:** normal
**Severity:** minor
**Version:** 4.04.0
**Fixed in version:** 4.05.0 +dev/beta1/beta2/beta3/rc1
**Category:** typing
**Monitored by:** @gasche @yallop

## Bug description

The exhaustivity/refutation checks do not taken into account the hidden _|_ member of the lazy type:

```ocaml
  # type (_, _) eq = Refl : ('a, 'a) eq;;
  type (_, _) eq = Refl : ('a, 'a) eq
  # type empty = (int, unit) eq;;
  type empty = (int, unit) eq
  # let f (x : ('a, empty Lazy.t) result) =
      match x with
      | Ok x -> x
      | Error (lazy _) -> .;;
        val f : ('a, empty Lazy.t) result -> 'a = <fun>
  # (f (Error (lazy (raise Not_found))) : string list list);;
  Segmentation fault
```

**Comment by vicuna** [comment]
**Comment author:** @garrigue

Fixed by commit 2608a18.
Do not split in a lazy pattern.

**Comment by vicuna** [comment]
**Comment author:** @gasche

Is this really fixed in 2608a18? It doesn't seem related to lazy patterns
  https://github.com/ocaml/ocaml/commit/2608a18

(I was curious to have a look at the patch. Would you have forgotten to commit it?)

**Comment by vicuna** [comment]
**Comment author:** @garrigue

Sorry, this was the fix for #7424.

This PR is actually fixed by commit d8429e5.
